File: myhdl/blink.py
#
# Copyright: 2013, Technical University of Denmark, DTU Compute
# Author: Martin Schoeberl (martin@jopdesign.com)
# License: Simplified BSD License
#

# Hardware Hello World

from myhdl import *

def blink(clk, led):

  CLK_FREQU = 16000000
  BLINK_FREQU = 1
  CNT_MAX = int(CLK_FREQU/BLINK_FREQU/2-1)
  
  cnt = Signal(intbv(0, min=0, max=CNT_MAX+1))
  blk = Signal(False)

  @always(clk.posedge)
  def hdl():
    if cnt == CNT_MAX:
      cnt.next = 0
      blk.next = not blk
    else:
      cnt.next = cnt+1

  @always_comb
  def comb():
    led.next = blk

  return hdl, comb

def ClkDriver(clk):

	interval = delay(1)
	@always(interval)
	def driveClk():
		clk.next = not clk

	return driveClk

clk = Signal(False)
led = Signal(False)
clk_driver = ClkDriver(clk)
inst = blink(clk, led)
sim = Simulation(clk_driver, inst)
sim.run(100)
toVHDL(blink, clk, led)

File: chisel/src/example/Blink.scala
/*
 * This code is part of the HDL comparison.
 * 
 * Copyright: 2013, Technical University of Denmark, DTU Compute
 * Author: Martin Schoeberl (martin@jopdesign.com)
 * License: Simplified BSD License
 * 
 * Blinking LED: the FPGA version of Hello World
 */

package example

import Chisel._
import Node._

class Blink() extends Module {
  val io = new Bundle {
    val led = UInt(OUTPUT, 1)
  }
  // BeMicro has a 16 MHz clock
  val CNT_MAX = UInt(16000000/2-1);
  val r1 = Reg(init = UInt(0, 25))
  val blk = Reg(init = UInt(0, 1))
  r1 := r1 + UInt(1)  
  when (r1 === CNT_MAX) {
    r1 := UInt(0)
    blk := ~blk
  }
  io.led := blk
}

// The sbt build tool passes some arguments to our main,
// which gives some arguments for the Chisel code generation

object BlinkMain {
  def main(args: Array[String]): Unit = {
    chiselMain(args, () => Module(new Blink()));
  }
}
